# 에러를 처리하는 여러가지 방법
21.02.15

mood: 😭

<! 공부하며 작성한 내용이며 틀린 내용이 있을 수 있습니다. >

## 오늘 한 것

- mmim과 가위바위보 순서도 작성 
- mmim Step1 PR 
- 에러 핸들링 개인 공부 

PR 작성하다가 날라가서, 다시 작성해야 했다. 토 할뻔...🤮

---

## 문제점 / 고민한점

* 에러 핸들링 
* 클래스와 구조체의 
* enum 타입을 어떻게 활용하는 것이 좋을까? 

### 에러 핸들링 방법 중 1번 
에러가 발생한 함수에서 리턴값으로 에러를 반환해 해당 함수를 호출한 코드에서 에러를 처리하도록 하는 방법 
do-catch 문을 사용하지 않고 어떻게 에러를 처리해야 하는지 모르겠다... 
해당 함수를 호출한 코드에서 어떻게 에러를 처리하도록 하는 것 일까? 

이에 대한 답변으로는 무조건 do-catch 문으로 받아서 처리를 해야 한다는 것이다. 


## 학습 내용

### 에러 핸들링 
Swift에서는 런타임에 에러가 발생한 경우 
그것의 처리를 위해 

- 에러의 발생(throwing),
- 감지(catching),
- 증식(propagating),
- 조작(manipulating)

을 지원하는 일급 클래스(first-class)를 제공합니다.

rror 프로토콜은 사실상 비어 있는 프로토콜이다.

Error 라는 프로토콜을 따르는 값이 에러처리를 위해 사용될 것임을 나타낼 뿐이다. 

Swift에서 4가지 에러를 처리 방법이 있다. 

1. 에러가 발생한 함수에서 리턴값으로 에러를 반환해 해당 함수를 호출한 코드에서 에러를 처리하도록 하는 방법 
2. do-catch 구문을 사용하는 방법
3. 옵셔널 값을 반환하는 방법
4. assert를 사용해 강제로 크래쉬를 발생시키는 방법입니다. 각각의 방법에 대한 설명은 이 후 섹션에 설명돼 있다.
---

(사실 1번은 2번과 같이 사용해야 한다. 에러를 발생시키고 난 후에 처리해 줄 곳이 필요하기 때문이다. )

3번은 우리가 알고 있는 옵셔널을 사용해서 처리하는 방법이다. (우리가 처리할 수 있는 에러일 경우엔 굳이 에러 메세지를 내지 않는 것도 방법이다.)

assert 를 사용하는 방법은 아직 공부하지 않았다.

📌 enum 타입으로 에러를 생성해 주기 
```swift
enum 에러 종류 이름: Error {
    case 에러 타입1
    case 에러 타입2
    case 에러 타입3
}
```
📌 에러를 발생시키기
`throw` 라는 키워드를 사용하면 에러를 발생시킬 수 있다.
```swift
throw 에러 종류 이름.에러 타입
```
throw 키워드를 만나면 컴파일러는 예상치 못한 상황이 발생했다는 것을 알라차린다.

📌 에러를 전파시키기 (throwing function)

함수 안에서 에러가 생긴다면,`throws` 라는 키워드를 붙여서 에러를 발생시킨다는 것을 알려주어야 한다. 

```swift
func canThrowErrors(input: Int) throws -> Int {
		guard input > 0 else {
		throw MyError.minorProblem
 }
```

이렇게 throw 가 구현부에 들어있어서 throws 를 붙여준 함수를 throwing function 이라고 한다. 

throwing function만이 에러를 발생시킬 수 있다. 

throwing function이 아닌 함수에서 throw가 발생한다면 반드시 그 함수내에서 throw에 대해 처리돼야 합니다. (do-catch 문이 내부에 있어야 함.)

guard 구문을 사용해 에러가 발생하면 함수에서 에러를 발생시키고 빠르게 함수를 탈출할 수 있도록 한다.(early exit)

이 메소드를 호출하는 메소드는 반드시 do-catch, try?, try! 등 구문을 앞에 붙여서 이 메서드가 오류를 밷을 수도 있다는 사실을 알려줘야 합니다.

```swift
try canThrowErrors(number)
```

📌 에러를 처리하기 (do-catch)

에러를 발생시켰으니 결국에는 에러를 처리해 줄 부분이 필요하다. 

```swift
do {
    try expression
    statements
} catch pattern 1 {
    statements
} catch pattern 2 where condition {
    statements
} catch {
    statements
}
```

함수를 `do-catch` 문 안에서 실행시키면 `catch` 라는 구문이 해당하는 오류를 잡고 { } 부분에서 어떻게 오류를 처리 해 줄 건지 처리를 해준다. 

## 추가적으로 에러를 처리해 주는 방법

try? 를 사용해서 리턴 값을 받아서 할당 해줄 수 도 있다. 

```swift
result = try? MyError.minorProblem
```

오류가 발생했을 때는 nil 을, 아닐 때는 옵셔널 벨류를 할당 해준다. 

---

선배들의 조언으로는 

제네릭 타입을 넣을 수 있는 Result  라는 것을 주로 사용하게 된다고 한다.  

그 부분도 제네릭을 확실히 익힌 후 공부해 보면 좋을 것 같다.



